lib/Sema/TypeCheckAttr.cpp

```diff
@@ -3679,6 +3860,237 @@ void AttributeChecker::visitDifferentiatingAttr(DifferentiatingAttr *attr) {
+  auto original = attr->getOriginal();
+  auto *transposeInterfaceType = transpose->getInterfaceType()
+                                     ->eraseDynamicSelfType()
+                                     ->castTo<AnyFunctionType>();
+  
+  // Get checked wrt param indices.
+  auto *wrtParamIndices = attr->getParameterIndexSubset();
+
+  // Get the parsed wrt param indices, which have not yet been checked.
+  // This is defined for parsed attributes.
+  auto parsedWrtParams = attr->getParsedParameters();
+  
+  bool wrtSelf = false;
+  if (!parsedWrtParams.empty())
+    wrtSelf = parsedWrtParams.front().getKind()
+                  == ParsedAutoDiffParameter::Kind::Self;
+
+  // If checked wrt param indices are not specified, compute them.
+  bool isCurried = transposeInterfaceType->getResult()->is<AnyFunctionType>();
+  if (!wrtParamIndices)
+    wrtParamIndices = computeTransposingParameters(
+                          TC, parsedWrtParams, transpose, isCurried,
+                          transpose->getGenericEnvironment(),
+                          attr->getLocation());
+  if (!wrtParamIndices) {
+    D->getAttrs().removeAttribute(attr);
+    attr->setInvalid();
+    return;
+  }
+
+  auto *expectedOriginalFnType =
+      transposeInterfaceType->getTransposeOriginalFunctionType(
+          attr, wrtParamIndices, wrtSelf);
+
+  // `R` result type must conform to `Differentiable`.
+  auto valueResultType = expectedOriginalFnType->getResult();
+  if (isCurried) {
+    valueResultType = transpose->mapTypeIntoContext(valueResultType->getAs<AnyFunctionType>()->getResult());
+  }
+  if (valueResultType->hasTypeParameter())
+    valueResultType = transpose->mapTypeIntoContext(valueResultType);
+  auto diffableProto = ctx.getProtocol(KnownProtocolKind::Differentiable);
+  auto valueResultConf = TC.conformsToProtocol(
+                             valueResultType, diffableProto,
+                             transpose->getDeclContext(), None);
+
+  if (!valueResultConf) {
+    TC.diagnose(attr->getLocation(),
+                diag::transposing_attr_result_value_not_differentiable,
+                expectedOriginalFnType);
+    D->getAttrs().removeAttribute(attr);
+    attr->setInvalid();
+    return;
+  }
+
+  // Compute expected original function type.
+  std::function<bool(GenericSignature *, GenericSignature *)>
+    checkGenericSignatureSatisfied =
+        [&](GenericSignature *source, GenericSignature *target) {
+          // If target is null, then its requirements are satisfied.
+          if (!target)
+            return true;
+          // If source is null but target is not null, then target's
+          // requirements are not satisfied.
+          if (!source)
+            return false;
+          // Check if target's requirements are satisfied by source.
+          // TODO: Assertion failure!
+          return TC.checkGenericArguments(
+                     transpose, original.Loc.getBaseNameLoc(),
```

Comment:
Please check the indentation here with `clang-format`. It should be a few spaces less.